### ISSUE TITLE
Bug: Unable to resolve "@logto/client/shim" from "node_modules/@logto/rn/lib/index.js"

### ISSUE THREAD

**Issue by alwyntan** [logto-io/react-native#20]
### Describe the bug
@logto/client/shim seems to not be exported from @logto/client and errors.

### Expected behavior
No crashes

### How to reproduce?
Install @logto/rn and import <LogtoProvider /> in the root component.

### Context
- [x] Logto Cloud
- [ ] Self-hosted, Logto version = 
  - [ ] Container (Docker image)
  - [ ] Raw Node.js

**Comment by gao-sun** [comment]
I'll take a look on it. It seems somehow the resolver doesn't honor the `exports` filed in `package.json`.

**Comment by gao-sun** [comment]
Which RN/Expo version are you using?

**Comment by alwyntan** [comment]
Using RN0.73.4 and expo 50.0.11

**Comment by shibilyt** [comment]
any progress in this? I am also facing the same issue.

**Comment by Mohamed-kassim** [comment]
same issue here expo sdk50 any progress?

**Comment by elvingomezdev** [comment]
Same problem, any update?

expo: 50.0.14

**Comment by JakimLi** [comment]
same issue, is this repo still being maintained? any update about this?

**Comment by flavindias** [comment]
same problem here, any updates?

**Comment by JakimLi** [comment]
I did some exploration, and finally solved this issue, this is not working because react native does not support package exports by default, I did below config to make it work:

in `metro.config.js `, create if not exist.
```js
const { getDefaultConfig } = require('expo/metro-config');  
// because I am using expo, so have to import expo's 
// default metro config, otherwise expo will complain with errors.

const config = getDefaultConfig(__dirname);

config.resolver.unstable_enablePackageExports = true;

module.exports = config;
```
this is the related react native doc: https://reactnative.dev/blog/2023/06/21/package-exports-support
hope this helps

**Comment by Housain-maina** [comment]
> I did some exploration, and finally solved this issue, this is not working because react native does not support package exports by default, I did below config to make it work:
> 
> in `metro.config.js `, create if not exist.
> 
> ```js
> const { getDefaultConfig } = require('expo/metro-config');  
> // because I am using expo, so have to import expo's 
> // default metro config, otherwise expo will complain with errors.
> 
> const config = getDefaultConfig(__dirname);
> 
> config.resolver.unstable_enablePackageExports = true;
> 
> module.exports = config;
> ```
> 
> this is the related react native doc: [reactnative.dev/blog/2023/06/21/package-exports-support](https://reactnative.dev/blog/2023/06/21/package-exports-support) hope this helps

This did the trick for me, thanks.

**Comment by Housain-maina** [comment]
> I did some exploration, and finally solved this issue, this is not working because react native does not support package exports by default, I did below config to make it work:
> 
> in `metro.config.js `, create if not exist.
> 
> ```js
> const { getDefaultConfig } = require('expo/metro-config');  
> // because I am using expo, so have to import expo's 
> // default metro config, otherwise expo will complain with errors.
> 
> const config = getDefaultConfig(__dirname);
> 
> config.resolver.unstable_enablePackageExports = true;
> 
> module.exports = config;
> ```
> 
> this is the related react native doc: [reactnative.dev/blog/2023/06/21/package-exports-support](https://reactnative.dev/blog/2023/06/21/package-exports-support) hope this helps

finally a fix, thanks

**Comment by gao-sun** [comment]
yes i'm working on the update. however i could not reproduce this issue. can you provide a repo for this issue?

**Comment by gao-sun** [comment]
i can successfully run and compile the 

> reactnative.dev/blog/2023/06/21/package-exports-support

@JakimLi thanks for providing the workaround. i'm curious of reproducing this issue because it went just fine on my laptop. even if i change `import` to `require` in `packages/rn-sample`.

**Comment by r3dm4n** [comment]
Adding this breaks axios for me.

**Comment by r3dm4n** [comment]
Is there any fix for this ?

**Comment by isener** [comment]
> I did some exploration, and finally solved this issue, this is not working because react native does not support package exports by default, I did below config to make it work:
> 
> in `metro.config.js `, create if not exist.
> 
> ```js
> const { getDefaultConfig } = require('expo/metro-config');  
> // because I am using expo, so have to import expo's 
> // default metro config, otherwise expo will complain with errors.
> 
> const config = getDefaultConfig(__dirname);
> 
> config.resolver.unstable_enablePackageExports = true;
> 
> module.exports = config;
> ```
> 
> this is the related react native doc: https://reactnative.dev/blog/2023/06/21/package-exports-support hope this helps

This has sorted the issue! 
Thanks a ton!!

**Comment by simeng-li** [comment]
Enabling packageExports in the Metro config is necessary to properly load @logto/client/shim. We've updated the integration documentation to include this information.

@isener, could you please share the detailed error information here?

It seems React Native needs to release a stable version that fully supports npm package exports.

**Comment by fvcoder** [comment]
> I did some exploration, and finally solved this issue, this is not working because react native does not support package exports by default, I did below config to make it work:
> 
> in `metro.config.js `, create if not exist.
> 
> ```js
> const { getDefaultConfig } = require('expo/metro-config');  
> // because I am using expo, so have to import expo's 
> // default metro config, otherwise expo will complain with errors.
> 
> const config = getDefaultConfig(__dirname);
> 
> config.resolver.unstable_enablePackageExports = true;
> 
> module.exports = config;
> ```
> 
> this is the related react native doc: https://reactnative.dev/blog/2023/06/21/package-exports-support hope this helps

Works

**Comment by dipakparmar** [comment]
> config.resolver.unstable_enablePackageExports = true;

Turning this on breaks other packages!

**Comment by simeng-li** [comment]
@dipakparmar can you try the latest 0.4.0 release. We have removed the import of the `shim` package.  `unstable_enablePackageExports` is no longer needed.

**Comment by dipakparmar** [comment]
> @dipakparmar can you try the latest 0.4.0 release. We have removed the import of the `shim` package. `unstable_enablePackageExports` is no longer needed.

Hey @simeng-li Thanks. Did the quick test and it works. Will do full test later this weekend and let you know. But so far its working.

**Comment by dipakparmar** [comment]
Confirming, this is resolved. Can successfully build the app on iOS.

**Comment by icodesign** [comment]
The new 0.4.0 release will break if we have `unstable_enablePackageExports` set to true, since [the PR](https://github.com/logto-io/js/pull/791) uses root `react-native` fields in package.json. 

While in office [React Native](https://reactnative.dev/blog/2023/06/21/package-exports-support#the-new-react-native-condition) and [metro](https://metrobundler.dev/docs/package-exports/#replacing-browser-and-react-native-fields) docs, enablePackageExports is going to be turned on by default in the future and the suggested way is to put `react-native` field under `exports` (instead of `root`). 

A compatible way is to add both (root and exports).
